File: test/spec/r/r_scripts/alpha.r
# equals?
(should (equals? true, true),
        (eq true))
(should (equals? true, false),
        (eq false))
(should (equals? true, ""),
        (eq false))

# equal
(should (equal true, false),
        (eq false))
(should (equal true, true, true),
        (eq true))
(should (equal true, false, true),
        (eq false))

# domain
(should (domain-of true),
        (eq Boolean))
(should (domain-of Alpha),
        (eq Domain))
(should (domain-of Boolean),
        (eq Domain))
(should (domain-of Domain),
        (eq Domain))
(should (domain-of "hello"),
        (eq String))
        
# is-a?
(should (is-a? Domain Integer)
        (eq true))
(should (is-a? Boolean true),
        (eq true))

File: lib/sbyc/r/system/operators/alpha.rb
SByC::R::Alpha::Operators.define{
  
  operator{|op|
    op.description = %Q{ Value equality }
    op.signature   = seq(system::Alpha, system::Alpha)
    op.argnames    = [:left, :right]
    op.returns     = system::Boolean
    op.aliases     = [:equals?, :==]
  }
  def equals?(left, right) left == right; end
  
  operator{|op|
    op.description = %Q{ Value equality (aggregate) }
    op.signature   = star(system::Alpha)
    op.argnames    = [:operands]
    op.returns     = system::Boolean
    op.aliases     = [:equal]
  }
  def equal(operands) 
    if operands.empty?
      true
    else
      first = operands.shift
      operands.all?{|op| equals?(first, op)}
    end
  end
  
  operator{|op|
    op.description = %Q{ Returns domain of a value }
    op.signature   = seq(system::Alpha)
    op.argnames    = [:operand]
    op.returns     = system::Domain
    op.aliases     = [:'domain-of']
  }
  def domain_of(operand) system::Alpha::domain_of(operand); end
  
  operator{|op|
    op.description = %Q{ Checks if a value belongs to a domain }
    op.signature   = seq(system::Alpha, system::Domain)
    op.argnames    = [:operand, :domain]
    op.returns     = system::Boolean
    op.aliases     = [:'is-a?']
  }
  def is_a?(operand, domain) domain.is_value?(operand); end
  
  operator{|op|
    op.description = %Q{ Coerces a value to a given domain }
    op.signature   = seq(system::Alpha, system::Domain)
    op.argnames    = [:operand, :domain]
    op.returns     = system::Alpha
    op.aliases     = [:coerce]
  }
  def coerce(operand, domain); domain.coerce(operand); end

  operator{|op|
    op.description = %Q{ Returns a literal for a given value }
    op.signature   = seq(system::Alpha)
    op.argnames    = [:operand]
    op.returns     = system::String
    op.aliases     = [:to_literal]
  }
  def to_literal(operand) domain_of(operand).to_literal(operand) end

}